src/main/kotlin/org/rust/lang/core/macros/ExpandedMacroStorage.kt

```diff
@@ -111,16 +117,26 @@ class ExpandedMacroStorage(val project: Project) {
         return stepped.asSequence().withIndex().map { (i, step) ->
             // Release memory. [map] contains prefetched calls only for the first step, so clear it on the second step
             if (i == 1) map.clear()
+            step?.map { sf -> Extractable(sf, workspaceOnly, map[sf]) }.orEmpty()
+        }
+    }
 
-            step?.map { sf ->
-                object : Extractable {
-                    override fun extract(): List<Pipeline.Stage1ResolveAndExpand> {
-                        val calls = if (sf.depth == 0) map[sf] else null
-                        return sf.extract(workspaceOnly, calls)
-                    }
-                }
-            } ?: emptyList()
+    fun removeSourceFilesFromStep(files: Collection<SourceFile>, step: Int) {
+        checkWriteAccessAllowed()
+        testAssert {
+            files.all { it.depth == step }
         }
+
+        stepped[step]?.removeAll(files)
+        files.forEach { it.depth = -1 }
```

Comment:
I've got such crash
```
java.lang.ArrayIndexOutOfBoundsException: Index -1 out of bounds for length 64
  at org.rust.lang.core.macros.ExpandedMacroStorage.getOrPutStagedList(ExpandedMacroStorage.kt:92)
  at org.rust.lang.core.macros.ExpandedMacroStorage.deserialize(ExpandedMacroStorage.kt:69)
  at org.rust.lang.core.macros.ExpandedMacroStorage.access$deserialize(ExpandedMacroStorage.kt:56)
  at org.rust.lang.core.macros.ExpandedMacroStorage$Companion.load(ExpandedMacroStorage.kt:257)
  at org.rust.lang.core.macros.MacroExpansionServiceImplInner$Companion.build(MacroExpansionManager.kt:441)
  at org.rust.lang.core.macros.MacroExpansionManagerImpl$projectOpened$1.call(MacroExpansionManager.kt:135)
  at org.rust.lang.core.macros.MacroExpansionManagerImpl$projectOpened$1.call(MacroExpansionManager.kt:106)
  at com.intellij.openapi.application.impl.ApplicationImpl$2.call(ApplicationImpl.java:274)
  at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
  at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  at java.base/java.lang.Thread.run(Thread.java:834)
```

b/c I restarted the IDE during macro expansion and it saved (and then loaded) `-1` as a depth of some SourceFile. Let's try to merge `removeSourceFilesFromStep` and `addSourceFilesToStep` into one write action in order to avoid any inconsistent state of the storage

Reply:
Looks ok now
Check it again, please